### PR TITLE
ARROW-7648: [C++] Sanitize local paths on Windows

### PR DESCRIPTION
**Pull request by pitrou**
FileSystemFromUri() now tries to convert backslashes to slashes if the original URI fails to parse.

### CODE REVIEW

**Comment by pitrou** [comment]
@fsaintjacques @jorisvandenbossche @kszucs Hopefully this will improve datasets on Windows.

**Comment by nealrichardson** [comment]
@ursabot crossbow submit conda-win-vs2015-py36 conda-win-vs2015-py37 conda-win-vs2015-py38

**Comment by kszucs** [comment]
[AMD64 Conda Crossbow Submit (#87669)](https://ci.ursalabs.org/#builders/98/builds/455) builder has been succeeded.

Revision: 31dcf0e2b5bd03594b7ecafe6b8fb3e0d470f336

Submitted crossbow builds: [ursa-labs/crossbow @ ursabot-447](https://github.com/ursa-labs/crossbow/branches/all?query=ursabot-447)

|Task|Status|
|----|------|
|conda-win-vs2015-py36|[Azure](https://dev.azure.com/ursa-labs/crossbow/_build/latest?definitionId=1&branchName=ursabot-447-azure-conda-win-vs2015-py36)|
|conda-win-vs2015-py37|[Azure](https://dev.azure.com/ursa-labs/crossbow/_build/latest?definitionId=1&branchName=ursabot-447-azure-conda-win-vs2015-py37)|
|conda-win-vs2015-py38|[Azure](https://dev.azure.com/ursa-labs/crossbow/_build/latest?definitionId=1&branchName=ursabot-447-azure-conda-win-vs2015-py38)|

**Comment by jorisvandenbossche** [comment]
This allows to pass a path with backslashes to `FileSystemFromUri`, but if you pass such a path to eg `GetTargetStats` that will still fail / produce paths with backslashes?

**Comment by nealrichardson** [comment]
Those nightly builds have been [failing](https://dev.azure.com/ursa-labs/crossbow/_build/results?buildId=5315&view=logs&j=4c86bc1b-1091-5192-4404-c74dfaad23e7&t=ec99a26b-0264-5e86-36fb-9cfd0ca0f9f3&l=4051), just want to see if this fixes that issue or if it's separate.

**Comment by pitrou** [comment]
> but if you pass such a path to eg GetTargetStats that will still fail / produce paths with backslashes?

Yes.

**Comment by jorisvandenbossche** [comment]
> > but if you pass such a path to eg GetTargetStats that will still fail / produce paths with backslashes?
>
> Yes.

Just to understand: is there a fundamental reason to not allow it there as well, or is it just "didn't tackle that here"? 

Currently, in the python datasets code, I am not passing each path through `FilesystemFromUri`, so ideally a path passed to eg `GetTargetStats` would also be sanitized IMO.

**Comment by pitrou** [comment]
I would turn the question on its head: why is it that you path non-abstract paths to `GetTargetStats`? Where do these paths come from?

**Comment by jorisvandenbossche** [comment]
> Where do these paths come from?

The user passes them? 
We could sanitize the strings on the python side though, with the exported `ToSlashes` from this PR?

**Comment by pitrou** [comment]
> The user passes them?

Do they? I thought dataset was doing filesystem-based partition discovery?

**Comment by pitrou** [comment]
AppVeyor build here: https://ci.appveyor.com/project/pitrou/arrow/builds/30288088

**Comment by jorisvandenbossche** [comment]
@pitrou rebase this now the windows tests PR is merged (https://github.com/apache/arrow/pull/6260)?

**Comment by pitrou** [comment]
Ok, rebased.

**Comment by pitrou** [comment]
AppVeyor build: https://ci.appveyor.com/project/pitrou/arrow/builds/30312131

File: cpp/src/arrow/filesystem/localfs_test.cc [hunk 1]
```diff
@@ -160,6 +187,32 @@ TYPED_TEST(TestLocalFS, FileSystemFromUriNoScheme) {
   this->TestFileSystemFromUri(this->local_path_);
 }
 
+TYPED_TEST(TestLocalFS, FileSystemFromUriFileBackslashes) {
+#ifdef _WIN32
+  this->TestFileSystemFromUri(ToBackslashes("file:" + this->local_path_));
+
+  // Variations
+  this->TestLocalUri("file:" + this->path_formatter_("C:foo\\bar"), "C:foo/bar");
```
Comment (bkietz) [inline cpp/src/arrow/filesystem/localfs_test.cc]: NB: `c:foo\bar` is a *relative* path, is this intended?

**Review by bkietz — CHANGES_REQUESTED** [review]
I think a neater way to handle this would be to require that the argument to `FileSystemFromUri` be rfc3986 formatted. The current fallback to localfs can be recovered by adding `static string Uri::FromPath(string)` which wraps [uriWindowsFilenameToUriStringA](https://uriparser.github.io/doc/api/latest/Uri_8h.html#a422dc4a2b979ad380a4dfe007e3de845) and the corresponding unix path function:
```c++
FileSystemFromUri(Uri::FromPath(R"(E:\dir\file.txt)"), &sanitized)
```
This is a little more boilerplate but I think it's worthwhile to be explicit here.

@jorisvandenbossche Since `FileSystemFromUri` accepts paths, sanitization can be achieved by discarding the filesystem and using the out path:
```c++
std::string sanitized;
RETURN_NOT_OK(FileSystemFromUri(Uri::FromPath(path), &sanitized).status());
```

**Comment by nealrichardson** [comment]
@bkietz could your feedback be done in a followup? This seems to unbreak the builds and a few things are depending on a solution to this.

**Comment by nealrichardson** [comment]
Appveyor on the fork just passed so this is as green as it is going to get today, given the backup on ASF appveyor.

File: cpp/src/arrow/filesystem/localfs_test.cc (hunk 1 above)
Comment (pitrou) [reply cpp/src/arrow/filesystem/localfs_test.cc]: Yes.

**Comment by pitrou** [comment]
@bkietz I didn't know those APIs. A solution is still needed so that the user can pass a path _or_ a URI to the same API. Whether this is done in `FileSystemFromUri` or in another higher-level function is a matter of discussion.

**Review by bkietz — APPROVED** [review]
Follow up JIRA for refactoring https://issues.apache.org/jira/browse/ARROW-7664

**Pull request closed without merging**